### ISSUE TITLE
Retrieving the index of the current linux repo causes "Invalid data in index" Exception

### ISSUE THREAD

**Issue by zpete** [libgit2/libgit2sharp#931]
Retrieving the index of the current linux repo (commit 53448a538d149dd9c66d09d4a7b416551d42b608) causes the following exception:

 `Invalid data in index - invalid entry`

Stacktrace:

```
 at LibGit2Sharp.Core.Ensure.HandleError(Int32 result) in p:\contrib\libgit2sharp\LibGit2Sharp\Core\Ensure.cs:line 128
   at LibGit2Sharp.Core.Ensure.ZeroResult(Int32 result) in p:\contrib\libgit2sharp\LibGit2Sharp\Core\Ensure.cs:line 146
   at LibGit2Sharp.Core.Proxy.git_repository_index(RepositorySafeHandle repo) in p:\contrib\libgit2sharp\LibGit2Sharp\Core\Proxy.cs:line 2191
   at LibGit2Sharp.Index..ctor(Repository repo) in p:\contrib\libgit2sharp\LibGit2Sharp\Index.cs:line 34
   at LibGit2Sharp.Repository.<.ctor>b__0() in p:\contrib\libgit2sharp\LibGit2Sharp\Repository.cs:line 62
   at LibGit2Sharp.Repository.<>c__DisplayClass11.<.ctor>b__2() in p:\contrib\libgit2sharp\LibGit2Sharp\Repository.cs:line 101
   at System.Lazy`1.CreateValue()
```

Reproducing the bug is pretty simple - just clone the current linux repo `https://github.com/torvalds/linux.git` and then try to retrieve the index:

```c#
using (var repo = new Repository("path/to/current/linux/repo")) 
{
    var index = repo.Index; // <-Exception "Invalid data in index - invalid entry"
}
```

Looks like the result of `NativeMethods.git_repository_index(out handle, repo);` is -1

**Comment by nulltoken** [comment]
@zpete Good catch!

Successfully reproduced on Windows:

```
$ git clone https://github.com/torvalds/linux.git
$ cd linux
$ git read-tree 53448a538d149dd9c66d09d4a7b416551d42b608
```

Then accessing the `Index` indeed throws.

/cc @carlosmn @ethomson

**Comment by shiftkey** [comment]
I've had a user identify this with another repository in GitHub for Windows. Going to try and recreate this on my end so I can poke around.

**Comment by DoctorVanGogh** [comment]
I'd like to push this bug a bit - open since one year without any progress.

I've just run into this very same issue - tracked it down from Visual Studio 2015 not cooperating with one of my repositories to the Microsoft Git Provider to libgit2sharp/libgit2. This is as far upstream as I can diagnose the issue.

A simple test program in libgit2sharp
```csharp
using (var repo = new Repository(@"path\to\repo")) {
    var idx = repo.Index;
}
```
errors out.

The problem persists through 
```bash
rm .git/index
git reset
```
a clone of the repo does not remove the issue - index is still considered 'invalid' by libgit2 (and all upstream consumers).

Git bash works perfectly fine on the repository, it's just libgit2 that's not cooperating.

This is currently a nasty showstopper for me - if a fix is not easy to implement, at least a somewhat more explicit error message would help: Something like `Invalid data in index - Entry [foo] invalid`.

My repo is currently private, so I can't supply it as a test case - if needed I might be able to bastardize together a test case with just the index, but no commits or objects.

**Comment by DoctorVanGogh** [comment]
Okay, I went ahead and anonymized my repo/index a bit. Testcase:

1. Create empty git repository
2. Place https://github.com/DoctorVanGogh/libgit2demo/blob/master/index in new repo's `.git` folder
3. Try accessing new Repository/Repository's Index property

=> Exception

**Comment by carlosmn** [comment]
Putting that index file into an empty repo tells me the index itself is corrupted. git complains about it as well.

```
$ git ls-files -s
error: bad index file sha1 signature
fatal: index file corrupt
```

What kind of filename are you using? We used to refuse indices with paths which are invalid for the operating system you're running on, but that would be fixed with the 0.22 release of libgit2sharp.

**Comment by carlosmn** [comment]
We no longer error out if an entry in the index would not be a valid filename on Windows, so the original issue is resolved with the 0.22 release. @DoctorVanGogh if you have issues with Visual Studio, use the MSFT support channels, as each project has their own fork of the library. If it's a problem in the upstream project, they will raise it to us.
